instrumentation/http-client/src/OpenTelemetry/Instrumentation/HttpClient.hs

```diff
@@ -13,143 +15,144 @@
- socket will be retained indefinitely. You must be certain to call
- 'responseClose' on this response to free up resources.
-
- This function automatically performs any necessary redirects, as specified
- by the 'redirectCount' setting.
-
- When implementing a (reverse) proxy using this function or relating
- functions, it's wise to remove Transfer-Encoding:, Content-Length:,
- Content-Encoding: and Accept-Encoding: from request and response
- headers to be relayed.
--}
-responseOpen :: (MonadUnliftIO m, HasCallStack) => HttpClientInstrumentationConfig -> Client.Request -> Client.Manager -> m (Client.Response Client.BodyReader)
-responseOpen httpConf req man = do
-  tracer <- httpTracerProvider
-  inSpan'' tracer "responseOpen" (addAttributesToSpanArguments callerAttributes spanArgs) $ \_ -> do
-    ctxt <- getContext
-    req' <- instrumentRequest httpConf ctxt req
-    resp <- liftIO $ Client.responseOpen req' man
-    _ <- instrumentResponse httpConf ctxt resp
-    pure resp
+import Control.Exception (assert)
+import Control.Monad ((>=>))
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.Bifunctor (bimap)
+import qualified Data.CaseInsensitive as CI
+import Data.Foldable (fold)
+import Data.Function ((&))
+import qualified Data.HashMap.Strict as H
+import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.Int (Int64)
+import Data.Monoid (Endo (Endo, appEndo))
+import qualified Data.TLS.GHC as TLS
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import GHC.Stack (HasCallStack, withFrozenCallStack)
+import qualified Network.HTTP.Client as Orig
+import qualified Network.HTTP.Types.Status as HT
+import qualified OpenTelemetry.Attributes.Map as Otel
+import qualified OpenTelemetry.Context.ThreadLocal as Otel
+import qualified OpenTelemetry.Propagator as Otel
+import qualified OpenTelemetry.SemanticConventions as Otel
+import qualified OpenTelemetry.Trace.Core as Otel
+
+
+appendModifierToSettings :: (MonadIO m, HasCallStack) => Otel.TracerProvider -> Orig.ManagerSettings -> m Orig.ManagerSettings
+appendModifierToSettings tracerProvider settings = withFrozenCallStack $ liftIO $ do
+  let
+    tracer =
+      Otel.makeTracer
+        tracerProvider
+        $Otel.detectInstrumentationLibrary
+        Otel.tracerOptions
+    managerModifyRequest = Orig.managerModifyRequest settings
+    managerModifyResponse = Orig.managerModifyResponse settings
+  tls <- makeThreadLocalStorage
+  pure
+    settings
+      { Orig.managerModifyRequest = requestModifier tracer tls >=> managerModifyRequest
+      , Orig.managerModifyResponse = managerModifyResponse >=> responseModifier tls
+      }
+
+
+type ThreadLocalStorage = TLS.TLS (IORef (Maybe Otel.Span))
+
+
+requestModifier :: HasCallStack => Otel.Tracer -> ThreadLocalStorage -> Orig.Request -> IO Orig.Request
+requestModifier tracer tls request = do
+  spanRef <- TLS.getTLS tls
```

Comment:
So, there's some kind of important set of assumptions here about how requests and responses are handled. It looks to me like you're assuming that a response to a request will always be exactly the next response processed by the thread that sent the response. Is that true? It could do with some explanation as to why we think it is true if so!

(That assumption is important because it means we can use a thread-local storage to effectively identify a request: if a response comes back and there is something in the storage, it must be for the request that we sent.)

Reply:
That point is certainly very important. I had made that assumption unconsciously. I will check it out. Maybe I need to think of another way.